### ISSUE TITLE
Map not working on newer bots (Deebot X1, T10, ...)

### ISSUE THREAD

**Issue by xXDjMikeXx** [DeebotUniverse/Deebot-4-Home-Assistant#268]
### Checks

- [X] I have searched the existing issues and no issue is describing my issue
- [X] I have checked the [FAQ](https://deebot.readthedocs.io/home/faq/)
- [X] I have checked the [documentation](https://deebot.readthedocs.io/integrations/home-assistant)
- [X] I have installed the latest version

### The problem

Map is not shown properly with Deebot X1 (last Model 950 worked without issues)

<img width="500" alt="image" src="https://user-images.githubusercontent.com/20925297/200549699-92feedee-6b3a-4f5f-aabb-570736a42927.png">

### Diagnostics information

[deebot-0fbf64dfdbed3ff9aedda7dc22e10a23-Sigbert-84ec494accf775ef946082380388d0e4.json.txt](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/files/9960491/deebot-0fbf64dfdbed3ff9aedda7dc22e10a23-Sigbert-84ec494accf775ef946082380388d0e4.json.txt)

**Comment by edenhaus** [comment]
I have only the 950 and as student I cannot afford to buy me the new X1. Probably they have changed the command, how you can retrieve the map data.

If you have the technical skills to intercept and analyse the traffic please feel free to do it. Each contribution is welcome :)
Also you can buy me some beers on https://www.buymeacoffee.com/edenhaus, so I can buy me the X1 in the future and the support for it :)

**Comment by xXDjMikeXx** [comment]
Hmm okay, I wish I could help but I don't know how to do this.... I do not have skills for stuff like that haha

**Comment by Sildenafil99** [comment]
![mappa](https://user-images.githubusercontent.com/49089530/203330096-460d16f9-c577-4c82-a72b-528e440e6568.png)
Same issue with Ecovacs T10 model

**Comment by Nismonx** [comment]
yes, same on my X1..... don't think this is a HA issue as valetudo on my other vacuum is displaying well.

if pointed in the right direction, I don't mind snifing the traffic...

tia

**Comment by xrh0905** [comment]
Same on my X1 OMNI.
I can snifing the traffic if you want, I just need to know the right place to look at.

**Comment by edenhaus** [comment]
Capture the traffic between the robot and the app.
Then filter for map commands and analyse them.
Probably your models are using V2 commands, which are currently not implemented.

**Comment by Nismonx** [comment]
> Capture the traffic between the robot and the app. Then filter for map commands and analyse them. Probably your models are using V2 commands, which are currently not implemented.

I understand what you are saying, just not sure what tools to use to do so.

Also, not sure if this is relevant at this point, but I believe there are currently issues with either the deebot app or robot itself when it comes to the map integrity. My self and other have been experiencing issues where the maps appear to have a mind of its own in the official app! In the last 2 weeks I had to re-do quick map several times. Rooms often merge together other time the actual rooms changes size. Don't even get me started on the 3D Maps lol.

Again,  I'm just speculating based on my experience and the comments of others in the official app in Google play.

**Comment by arohl** [comment]
I too have a X1 omni and see the same issue. I note that it would seem that @mrbungle64 who commented above has fixed it for his ioBroker.ecovacs-deebot code.

**Comment by mrbungle64** [comment]
@arohl 

> I too have a X1 omni and see the same issue. I note that it would seem that @mrbungle64 who commented above has fixed it for his ioBroker.ecovacs-deebot code.

No, that's not the case. Please check the linked issue (again).

**Comment by arohl** [comment]
> @arohl
> 
> > I too have a X1 omni and see the same issue. I note that it would seem that @mrbungle64 who commented above has fixed it for his ioBroker.ecovacs-deebot code.
> 
> No, that's not the case. Please check the linked issue (again).

I just saw the image from @djalexz and that looked like it was working to me...

Just noticed that the email comment I got from GitHub used sascha rather than @mrbungle64 - so of course you would know!

**Comment by mrbungle64** [comment]
@arohl 

> > @arohl
> > > I too have a X1 omni and see the same issue. I note that it would seem that @mrbungle64 who commented above has fixed it for his ioBroker.ecovacs-deebot code.
> > 
> > 
> > No, that's not the case. Please check the linked issue (again).
> 
> I just saw the image from @djalexz and that looked like it was working to me but I must admit I couldn't follow what the problem actually is (other than it doesn't display properly) nor what the solution (if indeed there is one) is

That issue is about the cleaning images provided by the cleaning logs of the Deebot API.
In the ioBroker adapter there's a state named "cleaninglog.lastCleaningMapImageURL" where the URL to the last cleaning map image is stored.

The user had not made this clear in the description of the issue.
We had clarified this in a Discord conversation. After that I had adjusted the title.
I'll add some further information to the issue to avoid further confusion 😉

**Comment by Marow** [comment]
Thats what i mean in my issue long time ago. But i dont know what i can do to help to fix that.
https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/issues/247

**Comment by TeCHiScy** [comment]
I recently analyzed the data of minor_map command on my X1 robot, the discovery is:
1. on X1 robot, value in a piece (100x100 array) is not only 0x01 (floor), 0x02 (wall), 0x03 (carpet). value = 0x02 forms only part of the house outline.
2. the lower 5 bit seems to represent the "area index". I divided the map into 32+ areas in the ecovacs app, and draw a gray-scale picture of lower 5 bit values, it shows same area divides as in the app. 
3. the higher 3 bit seems to be always 011, which forms a union set (i.e, complete map) of all the areas.

**Comment by edenhaus** [comment]
@TeCHiScy Please feel free to open a PR to make the map working on the X1

**Comment by TeCHiScy** [comment]
> @TeCHiScy Please feel free to open a PR to make the map working on the X1

Yes, I'm willing to do that. Can u share a picture of the map that works fine on your robot? I'd like to do some comparasion, to verify the results.

**Comment by edenhaus** [comment]
Maybe you need only to set the color for your new values [here](https://github.com/DeebotUniverse/client.py/blob/e87fefb9b58ee0aede9bb49955e9debc2557a142/deebot_client/map.py#L45).
Otherwise you need to update probably the [GetMinorMap](https://github.com/DeebotUniverse/client.py/blob/e87fefb9b58ee0aede9bb49955e9debc2557a142/deebot_client/commands/map.py#L285) command.

My map look similar to https://deebot.readthedocs.io/assets/images/custom_vacuum_card.jpg (it's a old image but nothing really changed)

**Comment by MetroAffe** [comment]
Same problem here.
For comparison I've made screenshots of both maps.
In the App are two different types of "floor". Only the dark grey areas and normal walls are rendered by this Integration.
It looks like the developer of Deebot added a second layer with Room-floors and AI generated walls.

![IMG_1980](https://user-images.githubusercontent.com/61084462/231436468-111dfb24-c8d0-40e7-ad86-6b79293016aa.jpg)
<img width="491" alt="Bildschirm­foto 2023-04-12 um 12 46 27" src="https://user-images.githubusercontent.com/61084462/231436542-361e7906-82f9-4a90-a9cf-990f99cad103.png">

**Comment by edenhaus** [comment]
I don't have a X1 and as I have nothing to do with ecovacs, I also don't get support from them. I cannot afford to buy each model. You can buy me always a coffee at https://www.buymeacoffee.com/edenhaus, so I can buy me new models.

Otherwise someone with a X1 model needs to analyze the traffic :)

Without a X1 it is hard to guess, which commands have changed and therefore currently I cannot fix this issue and need help for you guys, which have the X1.
I know about the problem and the only solution is to analyze the traffic or buy for me the X1 :smile:

**Comment by MetroAffe** [comment]
Okay well I will try to do this. But I need some help.
Which packages are important? How to provoke the commands? Is the connection not encrypted?

It would be no problem to send you all details or the whole traffic which are needed if you tell me some more Information.
Please contact me preferably by email and in maybe in German (TU Vienna? :D)

**Comment by Nismonx** [comment]
> Okay well I will try to do this. But I need some help. Which packages are important? How to provoke the commands? Is the connection not encrypted?
> 
> It would be no problem to send you all details or the whole traffic which are needed if you tell me some more Information. Please contact me preferably by email and in maybe in German (TU Vienna? :D)

You may need something like "packet capture" app or "wireshark" in a laptop. This is as far as I can help sorry!

**Comment by mrbungle64** [comment]
@MetroAffe @Nismonx @edenhaus 

I'll try to find the time this weekend to analyze the traffic.

**Comment by MetroAffe** [comment]
Yes, of course. I already captured the traffic between the App and Ecovacs/Deebot Server. But I think its encrypted - not surprising …. So I have no idea how to get the Information which commands to check for the Map. And I dont have the time to read into the API of Ecovacs/Deebot

**Comment by edenhaus** [comment]
> Yes, of course. I already captured the traffic between the App and Ecovacs/Deebot Server. But I think its encrypted - not surprising …. So I have no idea how to get the Information which commands to check for the Map. And I dont have the time to read into the API of Ecovacs/Deebot

If you for example use mitmproxy and setup it as proxy of the device you want to capture the traffic, you will see the traffic unencrypted.

@MetroAffe I prefer to stay here and write in english so everyone can follow and understand it

**Comment by MetroAffe** [comment]
Okay did it. And its working. I‘ve made a screenshot - more as proof that its running. Unfortunately I had to black out a lot because I don't like to do it publicly.

On which packets would you like to look closely at? If you would like to have the whole capture you know how to contact me. Its hard work to blure all private keys to share it in public environment…..

**Comment by MetroAffe** [comment]
Maybe I figured out some new commands and in my opinion this could be the Problem:
getMapInfo_V2
getMapSet_V2
getPos_V2

These got triggered too:
getSpeed
getBattery
getMinorMap
getMapTrace
getInfo
getCachedMapInfo
getMapInfo
getMajorMap

**Comment by edenhaus** [comment]
All currently implemented map commands can be found [here](https://github.com/DeebotUniverse/client.py/blob/main/deebot_client/commands/map.py).
For sure all `V2` commands are missing, but you need to check also the other map related commands, if there was added/changed some properties.

**Comment by MetroAffe** [comment]
I‘m on it 😊
How are the pieces put together at the command getminormap?
The „pieceindex“ order in my case is:
19 - 20 - 21
27 - 28 - 29
35 - 36 - 37

**Comment by MetroAffe** [comment]
Response of getCachedMapInfo starting with index 3 even it is empty:
{
"backupId": "",
"built": 0,
"index": 3,
"isFastBuilding": 0,
"mid": ""
"name": ""
"status": 1
"using": 0
},
{
"backupId": "123456789",
"built": 1,
"index": 0,
"isFastBuilding": 2,
"mid": "988764321"
"name": "Zuhause"
"status": 0
"using": 1
},
{
"backupId": "",
"built": 0,
"index": 2,
"isFastBuilding": 0,
"mid": ""
"name": ""
"status": 1
"using": 0
},
{
"backupId": "",
"built": 0,
"index": 2,
"isFastBuilding": 0,
"mid": ""
"name": ""
"status": 1
"using": 0
}

**Comment by edenhaus** [comment]
Please give a look at the current implementation to understand how the commands interact in between. Explaining here each command and how they interact is to much. Some commands are also documented [here](https://deebot.readthedocs.io/advanced/data_types/json/commands/map/) but still work in progress...
Also all docs is a guess from our side as ecovacs is not sharing any information

**Comment by MetroAffe** [comment]
That’s impossible. I already took a look at both. It’s not enough to understand a few lines of the code. For troubleshooting its needed to overview the whole code.

For now I’m done.
To everyone else: Take a look at following:
I think the Problem could also be „getMapSubSet“. This Command got never triggered and „getMapSet_V2“ includes subsets as one String now.
If that could be the problem I'm at the end of my abilities

**Comment by edenhaus** [comment]
Than please share information about the new `getMapSet_V2` command.

In short the map works:
- Call `getCachedMapInfo` to see which map is currently used. Storing `mid` (map id) for further processing
- Call `getMapSet` which the `mid` and all available [getMapSet](https://github.com/DeebotUniverse/client.py/blob/main/deebot_client/events/map.py#L60) Maybe the X1 has a new Type. Please analyse `getMapSet_V2`
- Call `getMapSubSet` for all `msid` (map subset id) returned by `getMapSet`. `getMapSubSet`  returns the map in the property `value`. Probably `getMapSet_V2` has different properties, which need to analyzed.

Please also note that analyzing the traffic is not a 5 minutes job. You need to the check some commands and verify what has changed. Maybe also some reverse engineering of the android app could be necessary to understand how the new data must be processed. And for all of this you need a lot of time

**Comment by MetroAffe** [comment]
getMapSet_V2 type ar:
```json
{
   "ret": "ok",
   "resp": {
      "header": {
         "pri": 1,
         "tzm": 120,
         "ts": "1681634062023",
         "ver": "0.0.1",
         "fwVer": "1.5.9",
         "hwVer": "0.1.1",
         "wkVer": "0.1.54"
      },
      "body": {
         "code": 0,
         "msg": "ok",
         "data": {
            "type": "ar",
            "mid": "451359723",
            "msid": "2012069423",
            "batid": "dbjhli",
            "serial": 1,
            "index": 1,
            "subsets": "XQAABACdAQAAAC2WwEHwYhHYQTGecueIxnXIJ1eT5msc+tUDcwyVN/qN4VpHCZICfcK0XSbLbHa2rDVyXa6W83QaBll7wDe0DbcHD7V0w73VqcenLcwRGTtLfhGyQAZckNp/W8R97OF03u2I1EVf3L/KtrKY+H8yXgcH9C7o2I4GrQaStxFptirT8uekcfwBeOTGNzwAe/TSzNO3H2IynQomiQgiADy+9AMs02A=",
            "infoSize": 413
         }
      }
   },
   "id": "7fGU",
   "payloadType": "j"
}
```

getMapSet_V2 type vw:
```json
{
   "ret": "ok",
   "resp": {
      "header": {
         "pri": 1,
         "tzm": 120,
         "ts": "1681634069523",
         "ver": "0.0.1",
         "fwVer": "1.5.9",
         "hwVer": "0.1.1",
         "wkVer": "0.1.54"
      },
      "body": {
         "code": 0,
         "msg": "ok",
         "data": {
            "type": "vw",
            "mid": "451359723",
            "batid": "cpipli",
            "serial": 1,
            "index": 1,
            "subsets": "XQAABAACAAAAAC2XPAAAAA==",
            "infoSize": 2
         }
      }
   },
   "id": "HSoS",
   "payloadType": "j"
}
```
getMapSet_V2 type mw:
```json
{
   "ret": "ok",
   "resp": {
      "header": {
         "pri": 1,
         "tzm": 120,
         "ts": "1681634069523",
         "ver": "0.0.1",
         "fwVer": "1.5.9",
         "hwVer": "0.1.1",
         "wkVer": "0.1.54"
      },
      "body": {
         "code": 0,
         "msg": "ok",
         "data": {
            "type": "mw",
            "mid": "451359723",
            "batid": "cpipli",
            "serial": 1,
            "index": 1,
            "subsets": "XQAABAACAAAAAC2XPAAAAA==",
            "infoSize": 2
         }
      }
   },
   "id": "zR84",
   "payloadType": "j"
}
```

**Comment by MetroAffe** [comment]
Well its much easier to get a solution. The guys from iobroker have had the same problem - and maybe already a solution!
They are working on a Version which implemented the new V2 commands:
getMapImage:
https://github.com/mrbungle64/ioBroker.ecovacs-deebot/compare/master...feature/v2/GetMapImage

**Comment by Nismonx** [comment]
@MetroAffe the commits are over a year old!

I'm afraid the last time Sascha stepped in recent conversations here, he confirmed it still a work in progress.

**Comment by mrbungle64** [comment]
@MetroAffe 

> Well its much easier to get a solution. The guys from iobroker have had the same problem - and maybe already a solution!

Not really 😉 

> They are working on a Version which implemented the new V2 commands

I already implemented some of the V2 commands, but I still have the same problem with the map generation.
You posted a link to an outdated feature branch. I'll delete this branch now 😉 

On Friday I already posted that I'll try to find the time this weekend to analyze the traffic:
https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/issues/268#issuecomment-1509035728

I'll let you know when I've found a solution.

**Comment by MetroAffe** [comment]
uh ok oops. I actually just overlooked the fact that it was from you and posted not the correct link. I read about it in this thread:
https://forum.iobroker.net/topic/55189/ecovacs-deebot-adapter-informationen-und-praxistipps/42?_=1681653819687&lang=de

If its not working yet ... How is it this possible:
https://forum.iobroker.net/assets/uploads/files/1677221745333-ecovacs.png


But I think in that case the problem is in the best hands! :D
If I could help you in any way with my newbie skills don't hesitate to ask.
Thanks for your work!

**Comment by mrbungle64** [comment]
@MetroAffe 

> If its not working yet ... How is it this possible:
> https://forum.iobroker.net/assets/uploads/files/1677221745333-ecovacs.png

This is the image of the last cleaning from the cleaning log 😉

**Comment by Nismonx** [comment]
@MetroAffe
If you want the map to look a little more presentable in Home assistant you can use a gimmick like I did just to keep the wife happy 😂

I've created a picture card with a floor plan sw to draw a plan of my home then lot's of editing in windows paint and then over imposed the x1 map over it.

Picture card :

```
type: picture-elements
elements:
  - type: image
    entity: camera.shuky_live_map
    camera_image: camera.shuky_live_map
    style:
      top: 54.9%
      left: 48.5%
      width: 89%
image: /local/community/lovelace-valetudo-map-card/HomePlanBlue.png
aspect_ratio: 135%

``` 

The map now looks like this:

![Screenshot_20230416_204624_Home Assistant](https://user-images.githubusercontent.com/62077532/232338835-7870eaca-1c41-45de-9375-5d1c286eaa1d.jpg)


It's off topic but thought you'd like the suggestion.

Great job capturing that data by the way 👍

**Comment by MacrosorcH** [comment]
Hi everybody, any progress at all with fixing the map for X1? @mrbungle64 Did you ever find a solution?
Thank you all for the great work!

**Comment by mrbungle64** [comment]
@MacrosorcH 

> @mrbungle64 Did you ever find a solution?

Yes and no. The solution is mostly only a theoretical one so far...

I think the only thing missing from the map is the border lines (boundaries).
I already had a short conversation with @edenhaus about it some month ago.

Some of the results I've already got looked very similar to the map in the Ecovacs Home app.
But the boundary data comes from a different dataset, so I have to refactor my code for it, because drawing in the lines for the boundaries doesn't work reliably yet. Unfortunately I haven't had the time to do this yet...

**Comment by edenhaus** [comment]
Here is a short update on my side. I will refactor the library to allow different commands for different models.
After that significant refactoring, we can implement V2 commands and activate them for some models.

Only with your help can I implement the required changes, as I don't have your model.
And as always, I can give no ETA.

**Comment by StefanoGiu** [comment]
I have a deebot X1 Omni, if you need help...

**Comment by TruantRiccardo** [comment]
Hi, I have a t10 plus and I am experiencing the same problem, so if you need information also from this model I can get it for you, if you point me to what.

**Comment by maoz82** [comment]
I can also support in case you need help, I've a Deebot X1 omni.

**Comment by edenhaus** [comment]
The new ecovacs models are using different commands for the map, which are currently not implemented.
I currently have no time to analyze these new commands and fix the map. Feel free to capture all commands and analyze them and open a PR for adding support. Please see also https://github.com/MVladislav/bumper/blob/dev/docs/dev/Ubuntu_Android_Sniffing.md#extend-mitmproy, which should help to capture commands but I haven't tested it yet.

**Comment by lotablet** [comment]
Any update about this? 
I have Deebot X1e Omni version, and this is exacty my problem too.
![Screenshot_20231216-071721~2](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/assets/62808137/cd514af2-3ec9-45b8-aff5-de0721cec48f)

I don't have many skills, but I can provide any information you need. Just teach me a little bit 😁

**Comment by TeCHiScy** [comment]
> @MacrosorcH
> 
> > @mrbungle64 Did you ever find a solution?
> 
> Yes and no. The solution is mostly only a theoretical one so far...
> 
> I think the only thing missing from the map is the border lines (boundaries). I already had a short conversation with @edenhaus about it some month ago.
> 
> Some of the results I've already got looked very similar to the map in the Ecovacs Home app. But the boundary data comes from a different dataset, so I have to refactor my code for it, because drawing in the lines for the boundaries doesn't work reliably yet. Unfortunately I haven't had the time to do this yet...

Same as me, still wonder how to get the boundaries as shown in the Ecovacs app.

**Comment by edenhaus** [comment]
https://github.com/DeebotUniverse/client.py/pull/372  tires to fix the map issue but it still needs some work before it can be merged. Please be patient
